xds/src/main/java/io/grpc/xds/XdsNameResolver.java

```diff
@@ -16,62 +16,84 @@
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.Nullable;
 
 /**
  * A {@link NameResolver} for resolving gRPC target names with "xds-experimental" scheme.
  *
- * <p>The implementation is for load balancing alpha release only. No actual VHDS is involved. It
- * always returns a hard-coded service config that selects the xds_experimental LB policy with
- * round-robin as the child policy.
+ * <p>Resolving a gRPC target involves contacting the traffic director via xDS protocol to
+ * retrieve service information and produce a service config to the caller.
  *
  * @see XdsNameResolverProvider
  */
 final class XdsNameResolver extends NameResolver {
 
-  // TODO(chengyuanzhang): delete this later, it was a workaround for demo.
-  @NameResolver.ResolutionResultAttr
-  static final Attributes.Key<Node> XDS_NODE = Attributes.Key.create("xds-node");
-  // TODO(chengyuanzhang): delete this later, XdsClient (to be implemented) constructed here
-  //  should take channel credentials config. This workaround is passing channel credentials
-  //  config to xDS load balancer, which is doing the xDS RPC communication for now.
-  @NameResolver.ResolutionResultAttr
-  static final Attributes.Key<List<ChannelCreds>> XDS_CHANNEL_CREDS_LIST =
-      Attributes.Key.create("xds-channel-creds-list");
+  private static final AtomicReference<BootstrapInfo> bootstrapInfoRef = new AtomicReference<>();
 
   private final String authority;
+  private final String hostName;
+  private final int port;
+  private final ObjectPool<XdsClient> xdsClientPool;
   private final Bootstrapper bootstrapper;
 
-  XdsNameResolver(String name) {
-    this(name, Bootstrapper.getInstance());
-  }
+  @Nullable
+  private XdsClient xdsClient;
 
-  @VisibleForTesting
-  XdsNameResolver(String name, Bootstrapper bootstrapper) {
+  XdsNameResolver(
+      String name,
+      final Args args,
+      final BackoffPolicy.Provider backoffPolicyProvider,
+      final Supplier<Stopwatch> stopwatchSupplier,
+      final XdsChannelFactory channelFactory,
+      Bootstrapper bootstrapper) {
     URI nameUri = URI.create("//" + checkNotNull(name, "name"));
-    Preconditions.checkArgument(nameUri.getHost() != null, "Invalid hostname: %s", name);
+    checkArgument(nameUri.getHost() != null, "Invalid hostname: %s", name);
     authority =
-        Preconditions.checkNotNull(
-            nameUri.getAuthority(), "nameUri (%s) doesn't have an authority", nameUri);
-    this.bootstrapper = Preconditions.checkNotNull(bootstrapper, "bootstrapper");
+        checkNotNull(nameUri.getAuthority(), "nameUri (%s) doesn't have an authority", nameUri);
+    hostName = nameUri.getHost();
+    port = nameUri.getPort();  // -1 if not specified
+    this.bootstrapper = checkNotNull(bootstrapper, "bootstrapper");
+    XdsClientFactory xdsClientFactory = new XdsClientFactory() {
+      @Override
+      XdsClient createXdsClient() {
+        BootstrapInfo bootstrapInfo = bootstrapInfoRef.get();
```

Comment:
It feels very bad to have `bootstrapInfoRef` scattered in different places.  I would delay creating `xdsClientPool` until when you create `xdsClient`, so `xdsClientPool` and `xdsClient` will have the same life cycle. And you don't have to introduce `bootstrapInfoRef` and use it in two places.

Reply:
I don't think delaying the creation of a `Pool` until the first time of getting an object from it is a good practice. A `Pool` should be a container (controls the lifecycle of pooled objects) and its usage should be similar to `AtomicReference`. It should be there at the first place while the creation of instances it holds can be delayed.

Given the existing `RefCountedXdsClientObjectPool` API, it has no way to know how to create an `XdsClient` instance until reading bootstrap is done. That's why I thought the idea of using a pool for `XdsClient` is not appropriate as the pool does not know how to create its pooled objects until somewhere in the middle of `XdsNameResolver`'s execution. Using `bootstrapInfoRef` is a last resort workaround for this issue.

Reply:
`XdsNameResolver` is the original creator of `XdsClientRef`, so it is right thing to create `XdsClientRef` at anytime before `getObject()` is first needed. To me, `start()` is a very proper and natural place to create it. If it were an `AtomicReference`, it's even right and natural to create the object and it's atomic ref together right before setting the atomic ref to the resolved addresses attributes.

I don't see any reason the pool must be there at the first place. In current situation, it's very awkward to create the pool in the constructor.

Reply:
What is `XdsClientRef`? Reference of `XdsClient` instance?

Think about `ForkJoinPool`, for which pooled objects are `Thread`. When to create instances of pooled object is pool's internal implementation. However, the __pool__ itself should be there at the time the object that uses it was created. You don't wait until you need a thread to create the __pool__.

Reply:
You shouldn't create a pool with an illegal state. `getObject()` will fail until `start()` is called. There is no point to create a pool that is not ready to use.

Reply:
Yes, this is bad. That's why I feel like there is no good way to use the concept of `ObjectPool` perfectly.

Reply:
Changed to create the pool after reading bootstrap is done to avoid having illegal state.